axlearn/common/checkpointer_orbax_emergency.py

```diff
@@ -0,0 +1,814 @@
+                        _orbax_save_fn, wrapped_save_policy=local_save_policy
+                    ),
+                    max_to_keep=cfg.local_keep_last_n,
+                ),
+                persistent=oecp.PersistentCheckpointOptions(
+                    should_save_fn=functools.partial(
+                        _orbax_save_fn, wrapped_save_policy=save_policy
+                    ),
+                    max_to_keep=cfg.keep_last_n,
+                ),
+                replica_axis_index=cfg.replica_axis_index,
+                async_options=oecp.checkpoint_manager.AsyncOptions(
+                    timeout_secs=cfg.async_timeout_secs
+                ),
+                step_name_format=self._name_format,
+                cleanup_tmp_directories=True,
+                enable_async_checkpointing=True,
+            ),
+        )
+        return self._tensor_manager
+
+    def save(
+        self, *, step: int, state: Nested[Tensor], evaler_summaries: Optional[Dict[str, Any]] = None
+    ):
+        """See `BaseCheckpointer.save` for details."""
+        assert self._eval_summaries is None, self._eval_summaries
+        self._eval_summaries = copy.deepcopy(evaler_summaries or {})
+        self._reached_preemption = self._tensor_manager.reached_preemption(step)
+
+        start_t = time.perf_counter()
+        state_with_tensors = jax.tree.map(
+            lambda x: x if isinstance(x, (Tensor, TensorSpec)) else None, state
+        )
+        # Note that save() waits for prior serialization to finish.
+        self._non_tensor_manager.save(step=step, state=state)
+        self._get_tensor_manager(state_with_tensors).save(
+            step=step, args=ocp.args.PyTreeSave(item=state_with_tensors)
+        )
+        self._eval_summaries = None
+        if (time_diff := time.perf_counter() - start_t) > 0.5:
+            logging.info("In-mem ckpt blocking time is %fs.", time_diff)
+        if self._reached_preemption:
+            self.wait_until_finished()
+            raise SystemExit(f"Exiting after saving checkpoint at {step=} due to pre-emption.")
+
+    def restore(
+        self,
+        *,
+        step: Optional[int] = None,
+        state: Union[Nested[Tensor], Nested[TensorSpec]],
+    ) -> Tuple[Optional[int], Nested[Tensor]]:
+        """See `BaseCheckpointer.restore` for details."""
+        start_t = time.perf_counter()
+        cfg: OrbaxEmergencyCheckpointer.Config = self.config
+        state_with_tensors = jax.tree.map(
+            lambda x: x if isinstance(x, (Tensor, TensorSpec)) else None, state
+        )
+        tensor_manager = self._get_tensor_manager(state_with_tensors)
+        if step is None:
+            # Find the intersection of the checkpoint steps managed by tensor and non-tensor
+            # manager, and then use the latest step in the intersection for restore. `all_steps`
+            # from tensor manager contains both local and persistent checkpoints.
```

Comment:
Consider refactoring this logic to a separate function so that it can be tested directly?

Reply:
Done and added a test.